community/server/src/main/java/org/neo4j/server/rest/discovery/DiscoverableURIs.java

```diff
@@ -0,0 +1,127 @@
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Optional;
+import java.util.function.BiConsumer;
+
+import org.neo4j.graphdb.config.InvalidSettingException;
+import org.neo4j.graphdb.config.Setting;
+import org.neo4j.helpers.AdvertisedSocketAddress;
+import org.neo4j.helpers.collection.Pair;
+import org.neo4j.kernel.configuration.Config;
+import org.neo4j.kernel.configuration.ConnectorPortRegister;
+import org.neo4j.server.configuration.ServerSettings;
+
+/**
+ * Repository of URIs that the REST API publicly advertises at the root endpoint.
+ */
+public class DiscoverableURIs
+{
+    private final Collection<Pair<String, String>> relativeUris = new ArrayList<>();
+    private final Collection<Pair<String, URI>> absoluteUris = new ArrayList<>();
+
+    public static DiscoverableURIs defaults( Config config, ConnectorPortRegister portRegister )
+    {
+        DiscoverableURIs repo = new DiscoverableURIs();
+        repo.addRelative( "data", config.get( ServerSettings.rest_api_path ).getPath() + "/" );
+        repo.addRelative( "management", config.get( ServerSettings.management_api_path ).getPath() + "/" );
+
+        URI bolt = discoverableBoltUri( "bolt", config, ServerSettings.bolt_discoverable_address, portRegister );
+        if ( bolt != null )
+        {
+            repo.addAbsolute( "bolt", bolt );
+        }
+        return repo;
+    }
+
+    public DiscoverableURIs addRelative( String key, String uri )
+    {
+        relativeUris.add( Pair.pair( key, uri ) );
+        return this;
+    }
+
+    public DiscoverableURIs addAbsolute( String key, URI uri )
+    {
+        absoluteUris.add( Pair.pair( key, uri ) );
+        return this;
+    }
+
+    public static URI discoverableBoltUri( String scheme, Config config, Setting<URI> override,
+            ConnectorPortRegister connectorPortRegister )
+    {
+        // Note that this whole function would be much cleaner to implement as a default function for the
+        // bolt_discoverable_address setting; however the current config design makes it hard to do
+        // "find any bolt connector", it can only do "find exactly this config key".. Something to refactor
+        // when we refactor config API for 4.0.
+        if ( config.isConfigured( override ) )
+        {
+            return config.get( override );
+        }
+
+        Optional<AdvertisedSocketAddress> boltAddress = config.enabledBoltConnectors().stream().findFirst().map(
+                boltConnector -> config.get( boltConnector.advertised_address ) );
+
+        if ( boltAddress.isPresent() )
+        {
+            AdvertisedSocketAddress advertisedSocketAddress = boltAddress.get();
+
+            // If port is 0 it's been assigned a random port from the OS, list this instead
+            if ( advertisedSocketAddress.getPort() == 0 )
+            {
+                int boltPort = connectorPortRegister.getLocalAddress( "bolt" ).getPort();
```

Comment:
What if bolt connector is named differently or there are multiple bolt connectors?